File: src/lib/software/foobar2000.py
from re import compile
from PySide6.QtCore import Slot
from src.lib.software import BaseSoftware, SoftwareCategory
from PySide6.QtNetwork import QNetworkReply, QNetworkRequest

class Foobar2000(BaseSoftware):
    def __init__(self):
        super().__init__()

        self.key = 'foobar2000'
        self.name = 'foobar2000'
        self.category = SoftwareCategory.Media
        self.download_name = 'foobar2000-x64.exe'
        self.is_archive = False
        self.should_cache_url = True
        self.requires_admin = False
        self.icon = 'foobar2000.png'
        self.homepage = 'https://foobar2000.org'

    @Slot(QNetworkReply)
    def on_manager_finished(self, reply: QNetworkReply):
        reply.deleteLater()
        error = reply.error()
        if error != QNetworkReply.NetworkError.NoError:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            return

        download_url_pattern = compile(r'foobar2000-x64_v\d+\.\d+\.\d+\.exe')
        html = reply.readAll().data().decode()
        match = download_url_pattern.search(html)
        if not match:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
        else:
            self.url_resolved.emit(f'https://foobar2000.org/files/{match.group(0)}')

    def resolve_download_url(self):
        self.manager.get(QNetworkRequest('https://www.foobar2000.org/download'))

File: packages/app/src/services/software/definitions/foobar2000.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory, DownloadUrlResolveError } from 'shared';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'foobar2000';
	public name = 'foobar2000';
	public category = [SoftwareCategory.Audio, SoftwareCategory.Media];
	public downloadName = 'foobar2000-x64.exe';
	public isArchive = false;
	public shouldCacheUrl = true;
	public requiresAdmin = false;
	public deprecated = false;
	public icon = 'foobar2000.png';
	public homepage = 'https://foobar2000.org';

	public async resolveDownloadUrl() {
		const res = await fetch('https://foobar2000.org/download');
		if (!res.ok) {
			return err(DownloadUrlResolveError.HTTPResponseError);
		}

		const html  = await res.text();
		const match = html.match(/foobar2000-x64_v\d+\.\d+\.\d+\.exe/);
		if (!match) {
			return err(DownloadUrlResolveError.Generic);
		}

		return ok(`https://foobar2000.org/files/${match[0]}`);
	}
}
